services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateTransformer.java

```diff
@@ -0,0 +1,280 @@
+                Vpc.VpcState vpcState = Vpc.VpcState.newBuilder().setConfiguration(vpcConfiguration).setOperationTypeValue(Common.OperationType.CREATE_VALUE).setOperationType(Common.OperationType.CREATE).build();
+                vpcStateSet.add(vpcState);
+            }
+
+            List<Router.RouterState> routerStateList = new ArrayList<Router.RouterState>();
+            if (networkConfiguration.getInternalRouterInfos() != null && neighborStates.keySet().stream().filter(e -> e.indexOf("#L3") != -1).count() > 0) {
+                for (InternalRouterInfo internalRouterInfo : networkConfiguration.getInternalRouterInfos()) {
+                    final List<InternalSubnetRoutingTable> subnetRoutingTables = internalRouterInfo.getRouterConfiguration().getSubnetRoutingTables();
+                    final List<Router.RouterConfiguration.SubnetRoutingTable> subnetRoutingTables2 = new ArrayList<Router.RouterConfiguration.SubnetRoutingTable>();
+                    for (InternalSubnetRoutingTable internalSubnetRoutingTable : subnetRoutingTables) {
+                        if (neighborStates.keySet().stream().filter(e -> e.indexOf("#L3") != -1).count() > 0) {
+                            final InternalSubnetEntity subnetEntity1 = subnetIdSubnetsMap.get(internalSubnetRoutingTable.getSubnetId());
+
+                            goalStateManager.getGoalStateHelper().add2SubnetStates(networkConfiguration, subnetStateSet, subnetEntity1);
+                        }
+
+                        Router.RouterConfiguration.SubnetRoutingTable subnetRoutingTable = Router.RouterConfiguration.SubnetRoutingTable.newBuilder().setSubnetId(internalSubnetRoutingTable.getSubnetId()).buildPartial();
+                        List<Router.RouterConfiguration.RoutingRule> routingRuleList = new ArrayList<Router.RouterConfiguration.RoutingRule>();
+                        for (InternalRoutingRule internalRoutingRule : internalSubnetRoutingTable.getRoutingRules()) {
+                            Router.DestinationType destinationType =
+                                    Router.DestinationType.INTERNET;
+                            Router.RouterConfiguration.RoutingRuleExtraInfo routingRuleExtraInfo = Router.RouterConfiguration.RoutingRuleExtraInfo.newBuilder().setDestinationType(destinationType).setNextHopMac(internalRoutingRule.getRoutingRuleExtraInfo().getNextHopMac()).build();
+                            Common.OperationType op = goalStateManager.getGoalStateHelper().getOperationType(internalRoutingRule.getOperationType().equals(OperationType.CREATE), internalRoutingRule.getOperationType().equals(OperationType.INFO), Common.OperationType.INFO, internalRoutingRule.getOperationType().equals(OperationType.DELETE), Common.OperationType.DELETE, internalRoutingRule.getOperationType().equals(OperationType.UPDATE), Common.OperationType.UPDATE);
```

Comment:
Can op = null? This is another potential NPE when trying to setOperationType(op) in next line.

Reply:
op is used internally and could not be null, but let me just change that per discussion

Reply:
addressed in 3acad53